### PR TITLE
fix(create-index): add debouncer for index name input and try to fix …

### PR DESCRIPTION
**Pull request by Ahton89**
…some ccs weighted requests

#### Description

This pr do is following:

1. Adds a debouncer for the input form for the name of the newly created index (timeout set to 700 ms)
2. Fixes calling useEffect 3 times due to obscure dependency indexPatterns.length
3. I'm not sure what I did right, but I tried to get rid of heavy requests to the elastic if it's CCS. I need the help of the community here, maybe I did it wrong. Added flag skipCCS in function getIndices.

this pr was an attempt to solve the following problem: when creating a new ccs index, the dashboard loaded the opensearch cluster very heavily. It turned out that in the first useEffect pulled 3 requests, the same requests were performed for each letter. Maybe my skipCCS is not needed, because the debouncer and fixing useEffect dependencies will not make many requests and the "opensearch" request will not load the opensearch cluster.

@ashwin-pc JFYI :)

### CODE REVIEW

**Comment by Ahton89** [comment]
And I fix useEffect 3 times execution on load page and when input index name, due to strange dependency of indexPattern.length

File: src/plugins/index_pattern_management/public/components/create_index_pattern_wizard/components/step_index_pattern/step_index_pattern.tsx [hunk 1]
```diff
@@ -39,6 +39,7 @@ import {
 } from '@elastic/eui';
 import { i18n } from '@osd/i18n';
 import { FormattedMessage } from '@osd/i18n/react';
+import _ from 'lodash';
```
Comment (ashwin-pc) [inline src/plugins/index_pattern_management/public/components/create_index_pattern_wizard/components/step_index_pattern/step_index_pattern.tsx:42]: Nit: you can import just the debounce function here.

```suggestion
import { debounce } from 'lodash';
```

File: src/plugins/index_pattern_management/public/components/create_index_pattern_wizard/components/step_index_pattern/step_index_pattern.tsx [hunk 2]
```diff
@@ -220,7 +222,7 @@ export class StepIndexPattern extends Component<StepIndexPatternProps, StepIndex
         getIndices({
           http,
           getIndexTags,
-          pattern: `${query}*`,
+          pattern: `${query}`,
```
Comment (ashwin-pc) [inline src/plugins/index_pattern_management/public/components/create_index_pattern_wizard/components/step_index_pattern/step_index_pattern.tsx:225]: Why was this change necessary?

**Review by ashwin-pc — COMMENTED** [review]
Firstly, thanks for submitting this PR @Ahton89. Can you make a few changes to the PR so that it's in line with our review guidelines?

1. Can you list each of your changes in the PR description? I think you have 3 distinct changes here. Calling each one out as a bullet point helps us reviewers know what the change is about.
2. Can you add a changelog entry for this fix in CHANGELOG.MD?

As for the review, I'm going to let someone with more context on CCS review that change and verify the use effect changes. I've reviewed your debounce change and it looks good.

File: src/plugins/index_pattern_management/public/components/create_index_pattern_wizard/components/step_index_pattern/step_index_pattern.tsx (hunk 1 above)
Comment (Ahton89) [reply src/plugins/index_pattern_management/public/components/create_index_pattern_wizard/components/step_index_pattern/step_index_pattern.tsx:42]: thanks, I'll fix it. :)

File: src/plugins/index_pattern_management/public/components/create_index_pattern_wizard/components/step_index_pattern/step_index_pattern.tsx (hunk 2 above)
Comment (Ahton89) [reply src/plugins/index_pattern_management/public/components/create_index_pattern_wizard/components/step_index_pattern/step_index_pattern.tsx:225]: perhaps this is a mistake, the project is very huge and while I debugged the logic of this component, I could leave this error after debugging.

**Comment by Ahton89** [comment]
Corrected the comments and tried to make a correct description PR.

**Comment by joshuarrrr** [comment]
@Ahton89 There are a couple of snapshot tests that need to be updated, which you can do from the command line and push as a new commit:

Snapshot Summary
 › 2 snapshots failed from 1 test suite. Inspect your code changes or run `yarn run test:jest:ci:coverage -u` to update them.

**Comment by Ahton89** [comment]
> @Ahton89 There are a couple of snapshot tests that need to be updated, which you can do from the command line and push as a new commit:
> 
> Snapshot Summary
> 
> > 2 snapshots failed from 1 test suite. Inspect your code changes or run `yarn run test:jest:ci:coverage -u` to update them.

I did something, I hope it's right :) I have a poppy m1 and saw an error in the console that such an architecture was not found - perhaps it is not supported. In general, if it's not difficult, can you just pick up this bug with this bullet and try to fix it? There, in the component of creating indexes, everything is very bad, as it seems to me, it may need to be refactored very much, because there is a feeling that they were engaged in the residual principle and, as they say, "thank you for working" :)

**Comment by foxdalas** [comment]
Hi @ashwin-pc, what do you think about this PR? We've been having a lot of load on the clusters with cross-cluster search when someone tries to create an index-pattern through the opensearch dashboard. This PR aims to fix that issue.

File: src/plugins/index_pattern_management/public/components/create_index_pattern_wizard/components/step_index_pattern/step_index_pattern.tsx [hunk 3]
```diff
@@ -272,7 +274,7 @@ export class StepIndexPattern extends Component<StepIndexPatternProps, StepIndex
 
     this.lastQuery = query;
     this.setState({ query, showingIndexPatternQueryErrors: !!query.length });
-    this.fetchIndices(query);
+    this.fetchIndices();
```
Comment (ashwin-pc) [inline src/plugins/index_pattern_management/public/components/create_index_pattern_wizard/components/step_index_pattern/step_index_pattern.tsx:277]: wont the change to use the query value from state result in the previous value of the query being fetched? The change you made to debounce this function might be the reason you arent seeing it, but removing the debounce would cause that issue. Was there any reason you decided to use the query value from state instead of the way it was initially setup?

File: src/plugins/index_pattern_management/public/components/create_index_pattern_wizard/components/step_index_pattern/step_index_pattern.tsx [hunk 4]
```diff
@@ -141,6 +142,7 @@ export class StepIndexPattern extends Component<StepIndexPatternProps, StepIndex
       initialQuery || context.services.uiSettings.get(UI_SETTINGS.INDEXPATTERN_PLACEHOLDER);
     this.state.indexPatternName = indexPatternCreationType.getIndexPatternName();
     this.dataSrouceEnabled = context.services.dataSourceEnabled;
+    this.fetchIndices = debounce(this.fetchIndices.bind(this), 700) as any;
```
Comment (ashwin-pc) [inline src/plugins/index_pattern_management/public/components/create_index_pattern_wizard/components/step_index_pattern/step_index_pattern.tsx:145]: Out of curiosity, how heavy of a load are we talking? I like the debounce here, but i wonder if we can take the delay lower. Also instead of reassigning `fetchIndices` to itself, can you assign it to a new variable like `debouncedFetchIndices`

File: src/plugins/index_pattern_management/public/components/index_pattern_table/index_pattern_table.tsx [hunk 5]
```diff
@@ -135,13 +135,7 @@ export const IndexPatternTable = ({ canSave, history }: Props) => {
...
       setIndexPatterns(gettedIndexPatterns);
     })();
-  }, [
-    history.push,
-    indexPatterns.length,
-    indexPatternManagementStart,
-    uiSettings,
-    savedObjects.client,
-  ]);
+  }, [history.push, indexPatternManagementStart, uiSettings, savedObjects.client]);
```
Comment (ashwin-pc) [inline src/plugins/index_pattern_management/public/components/index_pattern_table/index_pattern_table.tsx:138]: Good catch.

File: src/plugins/index_pattern_management/public/components/index_pattern_table/index_pattern_table.tsx [hunk 6]
```diff
@@ -168,11 +162,12 @@ export const IndexPatternTable = ({ canSave, history }: Props) => {
...
         setIsLoadingSources(false);
       });
-      getIndices({ http, pattern: '*:*', searchClient }).then((dataSources) =>
+      getIndices({ http, pattern: '*:*', searchClient, skipCCS: true }).then((dataSources) =>
         setRemoteClustersExist(!!dataSources.filter(removeAliases).length)
       );
     }
-  }, [http, creationOptions, searchClient, dataSourceEnabled]);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
```
Comment (ashwin-pc) [inline src/plugins/index_pattern_management/public/components/index_pattern_table/index_pattern_table.tsx:170]: this change and skipCCS are the two changes that I dint have a lot of context on. @kristenTian What are your thoughts on this change?

**Review by ashwin-pc — COMMENTED** [review]
Hey @Ahton89, sorry this took a bit to get back to, we were busy wrapping up things for the 2.5 release. @kristenTian @zengyan-amazon I think you two have a lot more context about how the index pattern creation flow should work. Can you two take a look at this issue and the fix suggested by @Ahton89 here?

In the meantime @foxdalas @Ahton89 Ive added my comments to the PR.

File: src/plugins/index_pattern_management/public/components/index_pattern_table/index_pattern_table.tsx (hunk 5 above)
Comment (Ahton89) [reply src/plugins/index_pattern_management/public/components/index_pattern_table/index_pattern_table.tsx:138]: yes, due to the fact that useEffect had variable indexPatterns.length, in its dependencies, useEffect worked 3 times with 3 requests, one of which was a cross-cluster and quite heavy. This fixes doing useEffect more than once :)

**Comment by Ahton89** [comment]
Let me show you the problem in screenshots :)
here are 3 screenshots. The bottom line is: when you open the index patterns page, due to an incorrect dependency in the useffect, 3 times 3 queries are executed. And I haven't clicked on anything yet, just opened the index patterns page. Then I press the button to create an index - again the same queries are executed, but once at least. Further without a debouncer for each character input - again 3 requests. This loads the opensource cluster very heavily. And with my pull request, I tried to solve this problem. But as I wrote above - I'm not a front-end guru and maybe my pull request is not quite right at some points. Therefore, I ask smarter colleagues to look at this matter, look at whether it is necessary to pull so many requests at each step, refactor and correct evil bugs that lead to a heavy load on the opensource cluster. Creating an index looks like a simple operation - react only needs to get the list of indexes created and the list of indexes from the cluster that haven't been added to the dashboard yet. And here the same work is constantly happening at every step, as it seems to me. I hope I was able to explain more clearly :)
bug-1
bug-2
bug-3

File: src/plugins/index_pattern_management/public/components/index_pattern_table/index_pattern_table.tsx (hunk 6 above)
Comment (ashwin-pc) [reply src/plugins/index_pattern_management/public/components/index_pattern_table/index_pattern_table.tsx:170]: The additional context was really helpful @Ahton89. So i tried your changes and it looks like `indexPatterns.length` is not the reason for the repeated calls to the cluster, but rather your change here. The same error does not occur when multiple data sources plugin is enabled. @kristenTian @zengyan-amazon Why was this effect skipped when the datasources plugin is enabled?

Comment (kristenTian) [reply src/plugins/index_pattern_management/public/components/index_pattern_table/index_pattern_table.tsx:170]: It is due to `getIndices` should not happen in here with Multiple Data Sources, but rather after user have selected a data source, OSD should then try to load the indices.

**Comment by ahopp** [comment]
@aoguan1990 Can we take a look and see what it would take to take this over the finish line? 

Also, @seraphjiang do you think this would help downstream managed services? For example, given the benefits to debouncing for features like cross-cluster, I could imagine this helping a ton.

**Comment by Ahton89** [comment]
Guys, can I help you with something else?  any thoughts?  plans?  offers?  say at least something :) maybe you took the bug for yourself and completely decided to rewrite this component?  :)

**Comment by foxdalas** [comment]
@ahopp its really works for cross-cluster indexes patterns. 
I checked it.

File: src/plugins/index_pattern_management/public/components/create_index_pattern_wizard/components/step_index_pattern/step_index_pattern.tsx (hunk 4 above)
Comment (kavilla) [reply src/plugins/index_pattern_management/public/components/create_index_pattern_wizard/components/step_index_pattern/step_index_pattern.tsx:145]: +1, I can't see where else this is used by I don't see it be explicitly private.

**Comment by kavilla** [comment]
> Guys, can I help you with something else? any thoughts? plans? offers? say at least something :) maybe you took the bug for yourself and completely decided to rewrite this component? :)

This a good start but I do believe a slight refactor to utilize the debounce as @ashwin-pc is a great idea. I do see it being debounced but I'm not positive if I can't say Ashwin's point about the previous query being used is not the case. It looks to be working but I believe I saw sometimes the previous query fire. 

Also another improvement can be to make it configurable to be fine tuned by the system admin for example a config in `opensearch_dashboards.yml`. This can make the debounce configurable if debouncing at all. I'd imagine there other locations where debouncing would be great so having it configurable can help the application not hit the cluster so hard.

If you aren't able to carry through I can commit onto this PR if that is okay.

Thank you again! This is awesome!

**Comment by Ahton89** [comment]
Hello @kavilla! Thank you very much for paying attention to the found problem and to my not ideal pull request. As I said, I'm not exactly a front-end developer, so it will be very cool if you connect and help in solving the problem found, so that the fix finally reaches the release :)

**Comment by jackychanel** [comment]
Hi All ! 

Think it's very important point cause in production mode (multiple datacenter) CCS just doesn't work 

We have exactly the same problem running both dashboards and Opensearch clusters in v1.3.1.
While just going to index pattern view on Dashboards make remote clusters node (in proxy mode) raise CPU/load until remote node left the cluster ... 

Sounds like a Dashboards v1.3.1 bug as Kibana 7.10  works just fine on those Opensearch backend clusters when creating index patterns, without generating any load on remote clusters.

It will be great if this point could be take in account  :)

**Comment by serge-r** [comment]
Hi guys!

We have similar problems with CCS. It's great to see that there's a solution for this issue, thank you for your work!

Is there any way we can assist to expedite this fix and see it in the release?

**Comment by jackychanel** [comment]
> @ahopp its really works for cross-cluster indexes patterns. I checked it.

We also build a new image integrate the modifications (as we need it right now in production), it works like a charm !
Thank you for the work ! 
IMHO , It should be merged as it's resolve production issues using CCS, even if you can iterate later by refit it more deeply

File: src/plugins/index_pattern_management/public/components/create_index_pattern_wizard/components/step_index_pattern/step_index_pattern.tsx [hunk 7]
```diff
@@ -141,7 +142,8 @@ export class StepIndexPattern extends Component<StepIndexPatternProps, StepIndex
     this.state.query =
       initialQuery || context.services.uiSettings.get(UI_SETTINGS.INDEXPATTERN_PLACEHOLDER);
     this.state.indexPatternName = indexPatternCreationType.getIndexPatternName();
-    this.dataSourceEnabled = context.services.dataSourceEnabled;
+    this.dataSrouceEnabled = context.services.dataSourceEnabled;
```
Comment (serge-r) [inline src/plugins/index_pattern_management/public/components/create_index_pattern_wizard/components/step_index_pattern/step_index_pattern.tsx:145]: @Ahton89 typo here dataSrouce -> dataSource

Comment (Ahton89) [reply src/plugins/index_pattern_management/public/components/create_index_pattern_wizard/components/step_index_pattern/step_index_pattern.tsx]: fixed, thanks

**Comment by serge-r** [comment]
Hi guys!

Is any updates on this? I see that 2.11 already in release, is new version have this fix?

**Comment by foxdalas** [comment]
@kavilla Hi! Do you have any updates about a bug?

**Comment by dblock** [comment]
Closing as stalled. It would be great if someone could pick this up again, rebase and finish it.

[Catch All Triage, attendees [1](https://github.com/dblock), [2](https://github.com/reta), [3](https://github.com/lukas-vlcek/), [4](https://github.com/rafaelma), [5](https://github.com/krisfreedain), [6](https://github.com/gargrajat), [7](https://github.com/andrross)]
